### PR TITLE
Support delay() in for loops

### PR DESCRIPTION
**Pull request by jackvreeken**
~~It's a very hacky implementation, specifically to get Channel Flow to work. It still does not properly run however, and I have no clue why. I'll put this on hold for now, until I find some more time/energy to work on this.~~

- [ ] Wait for #118  to get merged, as this PR is on top of that one.

### CODE REVIEW

**Comment by jbaayen** [comment]
Works fine for me with a simple example.

File: src/pymoca/backends/casadi/generator.py [hunk 1]
```diff
@@ -248,14 +249,24 @@ def exitExpression(self, tree):
                     entries.append(sym)
             src = ca.vertcat(*entries)
         elif op == 'delay' and n_operands == 2:
+            # Delay can accept an arbitrary expression, whereas we only seem to support states?
```
Comment (jbaayen) [inline src/pymoca/backends/casadi/generator.py]: See #115

File: src/pymoca/backends/casadi/generator.py [hunk 2]
```diff
@@ -355,7 +366,18 @@ def exitForEquation(self, tree):
             indexed_symbols_full = []
             for k in indexed_symbols:
                 s = f.indexed_symbols[k]
-                orig_symbol = self.nodes[self.current_class][s.tree.name]
+                if k.name() in [x.name for x in self.model.delayed_states]:
```
Comment (jbaayen) [inline src/pymoca/backends/casadi/generator.py]: Use generator here?

File: src/pymoca/backends/casadi/model.py [hunk 3]
```diff
@@ -50,9 +50,11 @@ def from_dict(cls, d):
...
         return variable
 
-
-DelayedState = namedtuple('DelayedState', ['name', 'origin', 'delay'])
-
+class DelayedState:
+    def __init__(self, name, origin, delay):
+        self.name = name
+        self.origin = origin
+        self.delay = delay
```
Comment (jbaayen) [inline src/pymoca/backends/casadi/model.py]: Why this?

File: src/pymoca/tree.py [hunk 4]
```diff
@@ -10,6 +10,7 @@
 import sys
 from collections import OrderedDict
 from typing import Union
+import os
```
Comment (jbaayen) [inline src/pymoca/tree.py]: No need for this I suppose

File: src/pymoca/backends/casadi/model.py [hunk 5]
```diff
@@ -361,9 +363,19 @@ def simplify(self, options):
                             rows.append(cols)
                         symbols.append(old_var.symbol)
                         values.append(ca.vertcat(*[ca.horzcat(*[v.symbol for v in row]) for row in rows]))
-                        new_vars.extend(itertools.chain.from_iterable(rows))
+                        new_vars_extra = list(itertools.chain.from_iterable(rows))
+                        if l == 'inputs':
```
Comment (jbaayen) [inline src/pymoca/backends/casadi/model.py]: I don't understand. We can also delay regular states or even algebraic states.

File: src/pymoca/backends/casadi/generator.py [hunk 6]
```diff
@@ -355,7 +366,18 @@ def exitForEquation(self, tree):
...
+                    # We are missing a similarly shaped delayed symbol. Replace/make it.
```
Comment (jbaayen) [inline src/pymoca/backends/casadi/generator.py]: We should probably change things such that the symbol in `self.model.delayed_states` has the right dimension for the for loop.

File: src/pymoca/backends/casadi/model.py (hunk 3 above)
Comment (jackvreeken) [reply src/pymoca/backends/casadi/model.py]: Have to overwrite field. Not possible with tuple.

File: src/pymoca/backends/casadi/model.py (hunk 5 above)
Comment (jackvreeken) [reply src/pymoca/backends/casadi/model.py]: Yes, but the test case was with inputs

**Comment by jackvreeken** [comment]
Picked up further in #118

**Pull request closed without merging**